### DISCUSSION TITLE
Is it easy to implement ConvTranspose op?

### DISCUSSION THREAD

**Discussion by zimond** [webonnx/wonnx#70] [General]
After a week of studying wonnx and trying to run my model, I find that the only missing piece is ConvTranspose. I'm really neither familiar with the pytorch cuda implementation nor wgsl. So is it easy to implement this op? Any instructions or information / learning material is appreciated. Thanks

**Comment by zimond** [comment]
https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/vulkan/ops/TransposeConvolution2d.cpp

**Comment by haixuanTao** [comment]
So, it shouldn't be that hard. I'm not sure on which axis the transpose is applied on, but if it is only on the last axis. You should be able to use the template of the convolution: https://github.com/webonnx/wonnx/blob/master/wonnx/templates/pool/conv.wgsl , add the transpose builtin function https://www.w3.org/TR/WGSL/#matrix-builtin-functions, before applying the product sum and you should be good.

If the axis is a different axis than the last one. You only have to redefine the chunk size based on the transposition formula. If you don't already now, each array of n dimension is represented by one array concatenating each dimension. 

Don't hesitate to ask more questions :+1:

**Reply by 101arrowz** [reply]
This isn't actually how ConvTranspose is implemented. I was able to implement a 2D ConvTranspose shader in GLSL for ONNX Runtime Web [here](https://github.com/microsoft/onnxruntime/pull/11990) for anyone who wants to try implementing it. I would be willing to do so after WebGPU gets enabled without a flag in at least some browsers.

**Comment by haixuanTao** [comment]
Note that this might not be the most optimised way of implementing it.

If this operator is really used a lot in your model you might look at other conv template https://github.com/webonnx/wonnx/tree/master/wonnx/templates/pool where we use group_size, vectorisation and padding to reduce the computation time.

**Reply by zimond** [reply]
Thanks! I will definitely give it a try. By the way is there any roadmap about ops coverage in onnx spec?
